File: tui/ui/theme.go
package ui

import "github.com/charmbracelet/lipgloss"

// ThemeForName returns the built-in Theme for the given name, and whether it
// was found. Unrecognised names return (TokyoNightStorm, false) so callers
// can fall back gracefully without a nil check.
func ThemeForName(name string) (Theme, bool) {
	t, ok := builtinThemes[name]
	return t, ok
}

// builtinThemes maps config.yaml theme names to their Theme values. Names
// mirror the TS CLI theme names so the shared config file works in both tools.
var builtinThemes = map[string]Theme{
	"default":              TokyoNightStorm,
	"tokyo-night":          TokyoNight,
	"tokyo-night-storm":    TokyoNightStorm,
	"tokyo-night-moon":     TokyoNightMoon,
	"tokyo-night-day":      TokyoNightDay,
	"catppuccin-mocha":     CatppuccinMocha,
	"catppuccin-macchiato": CatppuccinMacchiato,
	"catppuccin-frappe":    CatppuccinFrappe,
	"catppuccin-latte":     CatppuccinLatte,
	"dracula":              Dracula,
	"high-contrast":        HighContrast,
}

// Theme defines the colour palette for the TUI. Each field maps to a
// semantic role used across all panels, so swapping the theme changes
// the entire UI consistently.
//
// The default is Tokyo Night Storm. Users can override this with their
// own palette (Catppuccin, Gruvbox, Dracula, etc.) via the playback
// config file's [theme] section.
type Theme struct {
	// Background is the base background colour for all panels.
	Background lipgloss.Color
	// Foreground is the default text colour.
	Foreground lipgloss.Color
	// Clip is the colour for audio clip blocks on the timeline.
	Clip lipgloss.Color
	// ClipSelected is the colour for the currently-selected clip.
	ClipSelected lipgloss.Color
	// Overlap is the colour for overlapping regions and error/no-go states.
	Overlap lipgloss.Color
	// Warning is for states that need attention — edited, partial, pending.
	Warning lipgloss.Color
	// Delta is the colour for showing pause adjustments and success/go states.
	Delta lipgloss.Color
	// Muted is for de-emphasised text (placeholders, hints, inactive items).
	Muted lipgloss.Color
	// Accent is for titles, headings, and highlighted UI elements.
	Accent lipgloss.Color
	// Ruler is for the timeline ruler marks and gridlines.
	Ruler lipgloss.Color
	// Border is for panel borders and dividers.
	Border lipgloss.Color
}

// TokyoNightStorm is the default theme, based on the Tokyo Night colour
// scheme's "Storm" variant. Hex values sourced from the canonical palette
// at https://github.com/folke/tokyonight.nvim.
var TokyoNightStorm = Theme{
	Background:   lipgloss.Color("#24283b"),
	Foreground:   lipgloss.Color("#c0caf5"),
	Clip:         lipgloss.Color("#7aa2f7"),
	ClipSelected: lipgloss.Color("#7dcfff"),
	Overlap:      lipgloss.Color("#f7768e"),
	Warning:      lipgloss.Color("#e0af68"),
	Delta:        lipgloss.Color("#9ece6a"),
	Muted:        lipgloss.Color("#565f89"),
	Accent:       lipgloss.Color("#bb9af7"),
	Ruler:        lipgloss.Color("#3b4261"),
	Border:       lipgloss.Color("#545c7e"),
}

// HighContrast is a theme designed for low-vision users. It uses pure
// white on black with bold, saturated accent colours that exceed WCAG
// AAA contrast ratios (7:1+) against the black background. Muted text
// uses a lighter grey than Tokyo Night to remain legible.
var HighContrast = Theme{
	Background:   lipgloss.Color("#000000"),
	Foreground:   lipgloss.Color("#ffffff"),
	Clip:         lipgloss.Color("#00bfff"), // bright cyan — 8.6:1 on black
	ClipSelected: lipgloss.Color("#ffffff"), // pure white — maximum contrast
	Overlap:      lipgloss.Color("#ff5555"), // bright red — 5.5:1 on black, bold compensates
	Warning:      lipgloss.Color("#ff6600"), // bright orange — 4.6:1 on black, bold compensates
	Delta:        lipgloss.Color("#50fa7b"), // bright green — 11.8:1 on black
	Muted:        lipgloss.Color("#aaaaaa"), // light grey — 7.5:1 on black (AAA)
	Accent:       lipgloss.Color("#ffff55"), // bright yellow — 18.3:1 on black
	Ruler:        lipgloss.Color("#666666"), // mid grey — 4.2:1 on black (decorative only)
	Border:       lipgloss.Color("#aaaaaa"), // light grey — matches muted for consistency
}

// ── Tokyo Night variants ──────────────────────────────────────────────────────
// Hex values sourced from https://github.com/folke/tokyonight.nvim

// TokyoNight is the standard dark variant of the Tokyo Night palette.
var TokyoNight = Theme{
	Background:   lipgloss.Color("#1a1b26"),
	Foreground:   lipgloss.Color("#c0caf5"),
	Clip:         lipgloss.Color("#7aa2f7"), // blue
	ClipSelected: lipgloss.Color("#7dcfff"), // cyan
	Overlap:      lipgloss.Color("#f7768e"), // red
	Warning:      lipgloss.Color("#e0af68"), // yellow
	Delta:        lipgloss.Color("#9ece6a"), // green
	Muted:        lipgloss.Color("#414868"), // comment
	Accent:       lipgloss.Color("#bb9af7"), // purple
	Ruler:        lipgloss.Color("#283457"), // dark blue-grey
	Border:       lipgloss.Color("#3d59a1"), // blue-grey
}

// TokyoNightMoon is the Moon variant — darkest of the Tokyo Night family.
var TokyoNightMoon = Theme{
	Background:   lipgloss.Color("#222436"),
	Foreground:   lipgloss.Color("#c8d3f5"),
	Clip:         lipgloss.Color("#82aaff"), // blue
	ClipSelected: lipgloss.Color("#86e1fc"), // cyan
	Overlap:      lipgloss.Color("#ff757f"), // red
	Warning:      lipgloss.Color("#ffc777"), // yellow
	Delta:        lipgloss.Color("#c3e88d"), // green
	Muted:        lipgloss.Color("#444a73"), // comment
	Accent:       lipgloss.Color("#c099ff"), // purple
	Ruler:        lipgloss.Color("#2f334d"), // dark blue-grey
	Border:       lipgloss.Color("#444a73"), // matches muted
}

// TokyoNightDay is the light variant of the Tokyo Night palette.
var TokyoNightDay = Theme{
	Background:   lipgloss.Color("#e1e2e7"),
	Foreground:   lipgloss.Color("#3760bf"),
	Clip:         lipgloss.Color("#2e7de9"), // blue
	ClipSelected: lipgloss.Color("#007197"), // cyan
	Overlap:      lipgloss.Color("#f52a65"), // red
	Warning:      lipgloss.Color("#8c6c3e"), // yellow-brown
	Delta:        lipgloss.Color("#587539"), // green
	Muted:        lipgloss.Color("#848cb5"), // comment
	Accent:       lipgloss.Color("#9854f1"), // purple
	Ruler:        lipgloss.Color("#c4c8da"), // light grey
	Border:       lipgloss.Color("#a8aecb"), // blue-grey
}

// ── Catppuccin variants ───────────────────────────────────────────────────────
// Hex values sourced from https://github.com/catppuccin/catppuccin

// CatppuccinMocha is the darkest Catppuccin flavour.
var CatppuccinMocha = Theme{
	Background:   lipgloss.Color("#1e1e2e"), // base
	Foreground:   lipgloss.Color("#cdd6f4"), // text
	Clip:         lipgloss.Color("#89b4fa"), // blue
	ClipSelected: lipgloss.Color("#74c7ec"), // sapphire
	Overlap:      lipgloss.Color("#f38ba8"), // red
	Warning:      lipgloss.Color("#fab387"), // peach
	Delta:        lipgloss.Color("#a6e3a1"), // green
	Muted:        lipgloss.Color("#6c7086"), // overlay0
	Accent:       lipgloss.Color("#cba6f7"), // mauve
	Ruler:        lipgloss.Color("#313244"), // surface0
	Border:       lipgloss.Color("#585b70"), // surface2
}

// CatppuccinMacchiato is the medium-dark Catppuccin flavour.
var CatppuccinMacchiato = Theme{
	Background:   lipgloss.Color("#24273a"), // base
	Foreground:   lipgloss.Color("#cad3f5"), // text
	Clip:         lipgloss.Color("#8aadf4"), // blue
	ClipSelected: lipgloss.Color("#7dc4e4"), // sapphire
	Overlap:      lipgloss.Color("#ed8796"), // red
	Warning:      lipgloss.Color("#f5a97f"), // peach
	Delta:        lipgloss.Color("#a6da95"), // green
	Muted:        lipgloss.Color("#6e738d"), // overlay0
	Accent:       lipgloss.Color("#c6a0f6"), // mauve
	Ruler:        lipgloss.Color("#363a4f"), // surface0
	Border:       lipgloss.Color("#5b6078"), // surface2
}

// CatppuccinFrappe is the medium Catppuccin flavour.
var CatppuccinFrappe = Theme{
	Background:   lipgloss.Color("#303446"), // base
	Foreground:   lipgloss.Color("#c6d0f5"), // text
	Clip:         lipgloss.Color("#8caaee"), // blue
	ClipSelected: lipgloss.Color("#85c1dc"), // sapphire
	Overlap:      lipgloss.Color("#e78284"), // red
	Warning:      lipgloss.Color("#ef9f76"), // peach
	Delta:        lipgloss.Color("#a6d189"), // green
	Muted:        lipgloss.Color("#737994"), // overlay0
	Accent:       lipgloss.Color("#ca9ee6"), // mauve
	Ruler:        lipgloss.Color("#414559"), // surface0
	Border:       lipgloss.Color("#626880"), // surface2
}

// CatppuccinLatte is the light Catppuccin flavour.
var CatppuccinLatte = Theme{
	Background:   lipgloss.Color("#eff1f5"), // base
	Foreground:   lipgloss.Color("#4c4f69"), // text
	Clip:         lipgloss.Color("#1e66f5"), // blue
	ClipSelected: lipgloss.Color("#209fb5"), // sapphire
	Overlap:      lipgloss.Color("#d20f39"), // red
	Warning:      lipgloss.Color("#fe640b"), // peach
	Delta:        lipgloss.Color("#40a02b"), // green
	Muted:        lipgloss.Color("#9ca0b0"), // overlay0
	Accent:       lipgloss.Color("#8839ef"), // mauve
	Ruler:        lipgloss.Color("#ccd0da"), // surface0
	Border:       lipgloss.Color("#bcc0cc"), // surface2
}

// ── Dracula ───────────────────────────────────────────────────────────────────
// Hex values sourced from https://draculatheme.com/contribute

// Dracula is the classic purple-tinted dark theme.
var Dracula = Theme{
	Background:   lipgloss.Color("#282a36"),
	Foreground:   lipgloss.Color("#f8f8f2"),
	Clip:         lipgloss.Color("#8be9fd"), // cyan
	ClipSelected: lipgloss.Color("#ffffff"), // white
	Overlap:      lipgloss.Color("#ff5555"), // red
	Warning:      lipgloss.Color("#ffb86c"), // orange
	Delta:        lipgloss.Color("#50fa7b"), // green
	Muted:        lipgloss.Color("#6272a4"), // comment
	Accent:       lipgloss.Color("#bd93f9"), // purple
	Ruler:        lipgloss.Color("#44475a"), // current line
	Border:       lipgloss.Color("#6272a4"), // comment
}

File: src/theme.ts
/**
 * @file theme.ts
 * @module playback/theme
 *
 * CLI output colour themes for the playback logger.
 *
 * Each theme defines six semantic colour roles as ANSI SGR code strings
 * (the digits between `\x1b[` and `m`). An empty string means "no colour"
 * (the default terminal colour is used).
 *
 * Named presets can be loaded by name via {@link themeForName}. Users can
 * overlay individual roles from a `theme.yaml` file alongside the preset by
 * calling {@link loadTheme}.
 */

import { readFileSync, existsSync } from 'node:fs';
import { parse } from 'yaml';
import * as v from 'valibot';

// ── Colour role type ──────────────────────────────────────────────────────────

/**
 * Semantic colour roles used by the CLI logger reporter.
 * Values are ANSI SGR code strings, e.g. `"32"` for green, `"1;33"` for
 * bold yellow. An empty string means no colour (terminal default).
 */
export interface CliTheme {
	/** ANSI code for accented / highlighted text. */
	accent: string;
	/** ANSI code for `logError` fatal errors. */
	error: string;
	/** ANSI code for `logInfo` progress messages. */
	info: string;
	/** ANSI code for de-emphasised text (verbose/muted). */
	muted: string;
	/** ANSI code for `logSuccess` completion lines. */
	success: string;
	/** ANSI code for `logWarn` warnings. */
	warn: string;
}

// ── Built-in presets ──────────────────────────────────────────────────────────

/**
 * All built-in theme names accepted by {@link themeForName}.
 */
export type ThemeName =
	| 'default'
	| 'tokyo-night'
	| 'tokyo-night-storm'
	| 'tokyo-night-moon'
	| 'tokyo-night-day'
	| 'catppuccin-mocha'
	| 'catppuccin-macchiato'
	| 'catppuccin-frappe'
	| 'catppuccin-latte'
	| 'dracula'
	| 'high-contrast';

/**
 * Converts a hex colour string (`#rrggbb`) to an ANSI 24-bit foreground SGR
 * code (`38;2;r;g;b`). Terminals that do not support truecolour will degrade
 * gracefully to their nearest 256-colour equivalent.
 * @param h - Hex colour string in `#rrggbb` format.
 * @returns ANSI 24-bit foreground SGR code string.
 */
function hex(h: string): string {
	const r = parseInt(h.slice(1, 3), 16);
	const g = parseInt(h.slice(3, 5), 16);
	const b = parseInt(h.slice(5, 7), 16);
	return `38;2;${r};${g};${b}`;
}

/**
 * Named CLI theme presets. Hex values sourced from each scheme's canonical
 * palette definitions.
 */
export const THEMES: Record<ThemeName, CliTheme> = {
	// ── Catppuccin Frappé ────────────────────────────────────────────────────
	// https://github.com/catppuccin/catppuccin (medium)
	'catppuccin-frappe': {
		accent:  hex('#ca9ee6'), // mauve
		error:   hex('#e78284'), // red
		info:    hex('#8caaee'), // blue
		muted:   hex('#737994'), // overlay0
		success: hex('#a6d189'), // green
		warn:    hex('#ef9f76'), // peach
	},

	// ── Catppuccin Latte ─────────────────────────────────────────────────────
	// https://github.com/catppuccin/catppuccin (light)
	'catppuccin-latte': {
		accent:  hex('#8839ef'), // mauve
		error:   hex('#d20f39'), // red
		info:    hex('#1e66f5'), // blue
		muted:   hex('#9ca0b0'), // overlay0
		success: hex('#40a02b'), // green
		warn:    hex('#fe640b'), // peach
	},

	// ── Catppuccin Macchiato ─────────────────────────────────────────────────
	// https://github.com/catppuccin/catppuccin (medium dark)
	'catppuccin-macchiato': {
		accent:  hex('#c6a0f6'), // mauve
		error:   hex('#ed8796'), // red
		info:    hex('#8aadf4'), // blue
		muted:   hex('#6e738d'), // overlay0
		success: hex('#a6da95'), // green
		warn:    hex('#f5a97f'), // peach
	},

	// ── Catppuccin Mocha ─────────────────────────────────────────────────────
	// https://github.com/catppuccin/catppuccin (darkest dark)
	'catppuccin-mocha': {
		accent:  hex('#cba6f7'), // mauve
		error:   hex('#f38ba8'), // red
		info:    hex('#89b4fa'), // blue
		muted:   hex('#6c7086'), // overlay0
		success: hex('#a6e3a1'), // green
		warn:    hex('#fab387'), // peach
	},

	// ── Default ──────────────────────────────────────────────────────────────
	'default': {
		accent: '',
		error: '',
		info: '',
		muted: '2',
		success: '',
		warn: '33',
	},

	// ── Dracula ──────────────────────────────────────────────────────────────
	// https://draculatheme.com/contribute (purple-tinted dark)
	'dracula': {
		accent:  hex('#bd93f9'), // purple
		error:   hex('#ff5555'), // red
		info:    hex('#8be9fd'), // cyan
		muted:   hex('#6272a4'), // comment
		success: hex('#50fa7b'), // green
		warn:    hex('#ffb86c'), // orange
	},

	// ── High Contrast ────────────────────────────────────────────────────────
	// WCAG AAA contrast ratios (7:1+) on typical dark terminals.
	// Matches the TUI high-contrast theme's colour intent.
	'high-contrast': {
		accent:  '1;33',   // bold bright yellow
		error:   '1;31',   // bold bright red
		info:    '1;37',   // bold white
		muted:   '2',      // dim
		success: '1;32',   // bold bright green
		warn:    '1;33',   // bold bright yellow
	},

	// ── Tokyo Night ──────────────────────────────────────────────────────────
	// https://github.com/folke/tokyonight.nvim (Night variant)
	'tokyo-night': {
		accent:  hex('#bb9af7'), // purple
		error:   hex('#f7768e'), // red
		info:    hex('#7aa2f7'), // blue
		muted:   hex('#414868'), // comment
		success: hex('#9ece6a'), // green
		warn:    hex('#e0af68'), // yellow
	},

	// ── Tokyo Night Day ──────────────────────────────────────────────────────
	// https://github.com/folke/tokyonight.nvim (Day variant — light)
	'tokyo-night-day': {
		accent:  hex('#9854f1'),
		error:   hex('#f52a65'),
		info:    hex('#2e7de9'),
		muted:   hex('#848cb5'),
		success: hex('#587539'),
		warn:    hex('#8c6c3e'),
	},

	// ── Tokyo Night Moon ─────────────────────────────────────────────────────
	// https://github.com/folke/tokyonight.nvim (Moon variant)
	'tokyo-night-moon': {
		accent:  hex('#c099ff'),
		error:   hex('#ff757f'),
		info:    hex('#82aaff'),
		muted:   hex('#444a73'),
		success: hex('#c3e88d'),
		warn:    hex('#ffc777'),
	},

	// ── Tokyo Night Storm ────────────────────────────────────────────────────
	// Slightly darker variant — matches the TUI default theme
	'tokyo-night-storm': {
		accent:  hex('#bb9af7'),
		error:   hex('#f7768e'),
		info:    hex('#7aa2f7'),
		muted:   hex('#565f89'),
		success: hex('#9ece6a'),
		warn:    hex('#e0af68'),
	},
};

// ── YAML override schema ──────────────────────────────────────────────────────

const CliThemeOverrideSchema = v.partial(
	v.object({
		accent:  v.string(),
		error:   v.string(),
		info:    v.string(),
		muted:   v.string(),
		success: v.string(),
		warn:    v.string(),
	})
);

// ── Public API ────────────────────────────────────────────────────────────────

/**
 * Returns the built-in theme for the given name, or `undefined` if not found.
 * Use this when you have a validated name (e.g. from the XDG config schema).
 * @param name - A {@link ThemeName} string, e.g. `'tokyo-night-storm'`.
 * @returns The matching {@link CliTheme}, or `undefined` if the name is unknown.
 */
export function themeForName(name: string): CliTheme | undefined {
	return THEMES[name as ThemeName];
}

/**
 * Loads the active CLI theme.
 *
 * Resolution order:
 * 1. Built-in preset for `name` (falls back to `'default'` if unknown).
 * 2. `theme.yaml` at `overridePath` — if it exists, its colour role values
 *    are merged on top of the preset so users can swap individual colours
 *    without redefining the full palette.
 * @param name - Built-in theme name (e.g. `'tokyo-night-storm'`).
 * @param overridePath - Optional path to a `theme.yaml` override file.
 * @returns The resolved {@link CliTheme}.
 */
export function loadTheme(name: string, overridePath?: string): CliTheme {
	const base: CliTheme = themeForName(name) ?? THEMES['default'];

	if (!overridePath || !existsSync(overridePath)) {
		return base;
	}

	try {
		const raw = readFileSync(overridePath, 'utf8');
		const parsed = parse(raw) as unknown;
		const result = v.safeParse(CliThemeOverrideSchema, parsed);
		if (result.success) {
			return { ...base, ...result.output };
		}
	} catch {
		// Malformed theme.yaml — silently fall back to the preset.
	}

	return base;
}
